File: frontend/src/pet/variants.ts
// The anti-boredom engine: weighted-random variant pools, per-variant cooldowns, and a
// repetition penalty.
//
// **Why this exists at all.** Loona ships 700-1000 expressions and reviewers still describe it
// going on a shelf, so variety is not a content problem — it is an engine problem. Vector layers
// five independent anti-repetition systems; three of them are cheap and are implemented here.
//
// **Where the boundary sits.** The *action* is server-authoritative — `jpet/` decides that a poke
// means "wiggle". Only the *variant* is chosen here, because which of several wiggles you get is
// presentation, not pet state. This surface never invents what the pet did.
//
// **The age-specific tension, deliberately resolved.** Preschoolers love repetition; they will
// ask for the burp a hundred times. So recency is suppressed WITHIN a gag — you get a different
// burp — and never applied to the gag itself. Nothing here can refuse an action.

export interface Variant {
  /** The action key the renderer plays (`rig.ts:ACTIONS`). */
  v: string;
  /** Relative selection weight before cooldowns are applied. */
  w: number;
  /** Milliseconds after this variant plays before it may be chosen again. */
  cd: number;
}

/** One entry per thing the child can trigger repeatedly. A pool of one is fine and is how an
 *  action opts out of variation. */
export const POOLS: Record<string, Variant[]> = {
  poke: [
    { v: "wiggle", w: 3, cd: 2500 },
    { v: "giggle", w: 3, cd: 2500 },
    { v: "boing", w: 2, cd: 4000 },
    { v: "blush", w: 1, cd: 9000 },
    { v: "sneeze", w: 1, cd: 20000 },
    // Rare on purpose. A child who sees something once in three weeks talks about it for a
    // month; Vector ships calendar-gated animations at very low weight for the same reason.
    { v: "hiccup", w: 0.4, cd: 45000 },
  ],
  fart: [
    { v: "fart", w: 3, cd: 0 },
    { v: "sneeze", w: 1, cd: 12000 },
  ],
  dance: [
    { v: "bop", w: 3, cd: 2000 },
    { v: "shimmy", w: 2, cd: 4000 },
    { v: "dance", w: 1, cd: 7000 },
  ],
};

/** Mutable selection memory. Kept outside the pool tables so the tables stay constant data the
 *  firmware can put in flash. */
export interface PoolMemory {
  /** `${pool}:${variant}` → the ms timestamp it last played. */
  lastPlayed: Record<string, number>;
  /** `pool` → when it last fired, for the repetition penalty. */
  lastFired: Record<string, number>;
  /** `pool` → how many times in the current rapid burst. */
  streak: Record<string, number>;
}

export function newMemory(): PoolMemory {
  return { lastPlayed: {}, lastFired: {}, streak: {} };
}

/** Choose a variant, honouring cooldowns. `rand` is injected so selection is testable and the
 *  firmware can use its own PRNG.
 *
 *  **It can never fail.** If every variant is cooling down, the full pool is used rather than
 *  returning nothing — a child who pokes the pet must always get a reaction, and a silent
 *  no-op is the one outcome that reads as "broken". */
export function pickVariant(
  pool: string,
  mem: PoolMemory,
  now: number,
  rand: () => number = Math.random,
): string {
  const opts = POOLS[pool];
  if (!opts || opts.length === 0) return pool;
  const live = opts.filter(
    (o) => now - (mem.lastPlayed[`${pool}:${o.v}`] ?? Number.NEGATIVE_INFINITY) > o.cd,
  );
  const from = live.length > 0 ? live : opts;
  const total = from.reduce((a, o) => a + o.w, 0);
  let r = rand() * total;
  for (const o of from) {
    r -= o.w;
    if (r <= 0) {
      mem.lastPlayed[`${pool}:${o.v}`] = now;
      return o.v;
    }
  }
  // Unreachable while every weight is positive; kept so a zero-weight table still answers.
  const last = from[from.length - 1] ?? opts[0];
  if (!last) return pool;
  mem.lastPlayed[`${pool}:${last.v}`] = now;
  return last.v;
}

/** How long a burst has to be quiet before the penalty resets. */
export const PENALTY_RESET_MS = 12_000;
/** The floor: a heavily-repeated action still moves, it just moves less. Never zero — a
 *  motionless response is indistinguishable from a broken one. */
export const PENALTY_FLOOR = 0.35;

/** Magnitude multiplier for this firing (Vector's `repetitionPenalty`). The tenth identical
 *  poke in a burst lands softer than the first; leave it alone for 12 s and it is fresh again. */
export function repetitionPenalty(pool: string, mem: PoolMemory, now: number): number {
  const gap = now - (mem.lastFired[pool] ?? Number.NEGATIVE_INFINITY);
  mem.lastFired[pool] = now;
  if (gap > PENALTY_RESET_MS) {
    mem.streak[pool] = 0;
    return 1;
  }
  mem.streak[pool] = (mem.streak[pool] ?? 0) + 1;
  return Math.max(PENALTY_FLOOR, 0.82 ** mem.streak[pool]);
}

File: firmware/main/variants.c
#include "variants.h"

#include "rig.h"

typedef struct {
    int action;
    int weight; /* tenths, so the table stays integer: 0.4 in the web pool is 4 here */
    uint32_t cooldown_ms;
} variant_t;

/* One entry per thing the child can trigger repeatedly. Transcribed from
   `frontend/src/pet/variants.ts:POOLS`. `hiccup` is rare ON PURPOSE: a child who sees
   something once in three weeks talks about it for a month, which is why Vector ships
   calendar-gated animations at very low weight. */
static const variant_t POKE[] = {
    {ACT_WIGGLE, 30, 2500},
    {ACT_GIGGLE, 30, 2500},
    {ACT_BOING, 20, 4000},
    {ACT_BLUSH, 10, 9000},
    {ACT_SNEEZE, 10, 20000},
    {ACT_HICCUP, 4, 45000},
};

typedef struct {
    const variant_t *v;
    int n;
} pool_def_t;

static const pool_def_t POOLS[POOL_COUNT] = {
    [POOL_POKE] = {POKE, (int)(sizeof(POKE) / sizeof(POKE[0]))},
};

/* A variant that has never played is never cooling, however early it is. */
static int cooling(const pool_def_t *def, const pool_memory_t *mem, int i, uint32_t now_ms)
{
    if ((mem->played_mask & ((uint32_t)1 << i)) == 0) return 0;
    return (uint32_t)(now_ms - mem->last_played[i]) <= def->v[i].cooldown_ms;
}

void variants_reset(pool_memory_t *mem)
{
    if (mem == NULL) return;
    for (unsigned i = 0; i < sizeof(mem->last_played) / sizeof(mem->last_played[0]); i++) {
        mem->last_played[i] = 0;
    }
    mem->last_fired = 0;
    mem->streak = 0;
    mem->played_mask = 0;
    mem->fired = 0;
}

int variants_pick(pool_t pool, pool_memory_t *mem, uint32_t now_ms, uint32_t rnd)
{
    if ((unsigned)pool >= (unsigned)POOL_COUNT) return ACT_WIGGLE;
    const pool_def_t *def = &POOLS[pool];
    if (def->n <= 0 || mem == NULL) return ACT_WIGGLE;

    /* Two passes over the same table: first only the variants off cooldown, and if that is
       empty, all of them. IT CAN NEVER FAIL — a poke that produces nothing is the one outcome
       a child reads as "broken", so a fully-cooling pool still answers. */
    for (int pass = 0; pass < 2; pass++) {
        int total = 0;
        for (int i = 0; i < def->n; i++) {
            if (pass == 0 && cooling(def, mem, i, now_ms)) continue;
            total += def->v[i].weight;
        }
        if (total <= 0) continue;
        int r = (int)(rnd % (uint32_t)total);
        for (int i = 0; i < def->n; i++) {
            if (pass == 0 && cooling(def, mem, i, now_ms)) continue;
            r -= def->v[i].weight;
            if (r < 0) {
                mem->last_played[i] = now_ms;
                mem->played_mask |= (uint32_t)1 << i;
                return def->v[i].action;
            }
        }
    }
    return def->v[0].action;
}

float variants_penalty(pool_t pool, pool_memory_t *mem, uint32_t now_ms)
{
    (void)pool;
    if (mem == NULL) return 1.0f;
    const uint32_t gap = now_ms - mem->last_fired;
    const int first = !mem->fired;
    mem->last_fired = now_ms;
    mem->fired = 1;
    if (first || gap > PENALTY_RESET_MS) {
        mem->streak = 0;
        return 1.0f;
    }
    mem->streak++;
    float mag = 1.0f;
    for (int i = 0; i < mem->streak && i < 32; i++) mag *= 0.82f;
    return mag < PENALTY_FLOOR ? PENALTY_FLOOR : mag;
}
